Guidelines emerged after code reviews (#18)
pagopa/io-handbook: development-styleguide.md

- Such practice should generally be avoided as it can lead to maintenance problems and can produce hard to find bugs
- If you don't own it, don't modify it

```typescript
let bob = { name: "Bob", surname: "Smith", age: 23 };
bob.age = 24; // avoid
// create a new object using spread operator to update only specific fields
const olderBob = { ...bob, age: 24 };
```

#### Avoid hiding exceptions

> Usually in Javascript and Typescript is hard to figure out if a function throws an exception. Try to be explicit about the exceptions you throw.

Reviewer: ```suggestion
Usually in Javascript and Typescript is hard to figure out whether a function throws an exception or not. Try to be explicit about the exceptions you throw.
```

> Usually in Javascript and Typescript is hard to figure out whether a function throws an exception or not. Try to be explicit about the exceptions you throw.

Reviewer 2: I'd add: "try not to use exceptions at all, you can use eithers instead in 99% of cases. Exceptions must be limited to the scope of the main() method in case you need to early exit the process with an error message"

```typescript
//
// ✅good: exception is thrown explicitly
//

const config = (key: string): Option<string> => (
  fromNullable(process.env[key])
);

> const callService () => {

Reviewer: even better would be to document the function, esp. if the function gets imported from a library: in Typescript jumping to the code from your IDE will get you to the type definitions file, not to the source, so you'll not be able to see the actual throws unless they're documented.

Author: Comments are liabilities ;)
Said that, I think we can add some principles about _practical code comments_, 

Back to the guideline, do you think we can improve this definition?

Reviewer: not sure, I would say to not throw exceptions altogether, just return an `Either`

Reviewer: regarding documentation as a liability, I agree on those cases where documentation doesn't add any value (i.e. it's redundant) - in all other cases, when documentation describes what is not obvious, the documentation is actually an asset and thus it needs to be maintained as part of the code base, see [this discussion](https://trello.com/c/1CzV7LPb) I've started.

Author: If we have some mechanism to make comments in sync with the actual behavior of the code, then we can trust it, of course. I leave this to the broader discussion you mentioned.

About exceptions: I think we have broader thoughts about this subject, I'll summarize in a comment below.

  }
);
```

if your validation function could raise multiple errors
you should use [Validation](https://dev.to/gcanti/getting-started-with-fp-ts-either-vs-validation-5eja) to collect them.

#### Avoid the use of imperative methods like `isLeft` and `isNone`.

When possible avoid to use `isNone` / `isLeft` checks that makes code verbose (and more imperative). Use functional construct from [`fp-ts`](https://gcanti.github.io/fp-ts/modules/Option.ts.html) types instead (fold, map, mapLeft, bimap, getOrElse, ...).

```typescript
//
// ✅good: use of fp-ts's algebric constructs
//

// with Option
const name: string = fromNullable(user.name).getOrElseL(() => {
  throw new Error("name required!");
});

// with Either
const validString: string = validatedStringEither.fold(
  err => {

>     throw new Error("invalid string!");

Reviewer 2: I'd avoid using exceptions into examples here :-)

  return { amount: prevState.amount + 1 };
});
```

### TESTING

> #### Prefer test titles in the form _"it should ..."_

Reviewer 2: this is a must

```typescript
// ✅good
it("should have `name` property", () => { ... });

// ❌bad: test title must be in "should" form
it("has `name` property", () => { ... });

// ❌bad: use "it" instead of "test"
test("should have `name` property", () => { ... });
```

> #### Avoid throw exceptions

Reviewer: ```suggestion
#### Avoid throwing exceptions
```

> Sometimes test code can have unhappy branches you don't want to test and hence consider as a test failure. Try not to throw exception in there.

Reviewer: ```suggestion
Sometimes test code can have unhappy branches you don't want to test and hence consider as a test failure. Try not to throw exceptions in there.
```

If you are using `Jest`, you can use [`expect.assertions`](https://jest-bot.github.io/jest/docs/expect.html#expectassertionsnumber).

```typescript
// ✅good
expect.assertions(1);
myDecoder.fold(
  () => {},
  (value) => {
    expect(value).toEqual( ... );
  }
);

// ❌bad
myDecoder.fold(
  () => {
    throw new Error("test should not pass here");
  },
  (value) => {
    expect(value).toEqual( ... );
  }
);

```

#### Avoid wrapping tests with helper functions

> Be explicit in the test you are executing. If you are repeating te same test in many places, don't be tempted to wrap common tests into helpers. We encourage code duplication in favor of more explicit code.

Reviewer: ```suggestion
Be explicit in the test you are executing. If you are repeating te same test in many places, don't be tempted to wrap common tests into helpers. We encourage code duplication in favor of more explicit logic.
```
